pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderImpl.java

```diff
@@ -177,17 +177,15 @@ public Message<T> readNext(int timeout, TimeUnit unit) throws PulsarClientExcept
 
     @Override
     public CompletableFuture<Message<T>> readNextAsync() {
-        CompletableFuture<Message<T>> receiveFuture = consumer.receiveAsync();
-        receiveFuture.whenComplete((msg, t) -> {
-           if (msg != null) {
-               consumer.acknowledgeCumulativeAsync(msg).exceptionally(ex -> {
-                   log.warn("[{}][{}] acknowledge message {} cumulative fail.", getTopic(),
-                           getConsumer().getSubscription(), msg.getMessageId(), ex);
-                   return null;
-               });
-           }
+        return consumer.receiveAsync().thenApply(msg -> {
```

Comment:
How does this fix work? Your error message says `msg` is not null but `msg.getMessageId` is null.

Unless I'm reading this wrong, your change is a cleaner way but the effect remains the same. Baring very unexpected behaviors (for example, `consumer.acknowledgeCumulativeAsync(msg)` throwing an exception, which before the change is silently disgarded.).

If `msg.messageId()` somehow is null, you would have the same error, no?

Reply:
Here, the modification is to return the CompetedFuture downstream from `thenApply(msg -> {acknowledgeCumulativeAsync ) ` instead of `receiveFuture` like `MultiTopicsReade.readNextAsync()`

The messageId can be null after `message.release()`.